examples/vnet/doca_gw_dpdk.c

```diff
@@ -68,18 +99,190 @@ int doca_gw_port_stop(struct doca_gw_port *port)
+		spec->hdr.dst_addr = match->in_dst_ip.a.ipv4_addr;
+	if ((param->flags & SET_OUTER_DST_IP) && match->out_dst_ip.a.ipv4_addr)
+		spec->hdr.dst_addr = match->out_dst_ip.a.ipv4_addr;
+	return 1;
+}
+
+static void build_ipv4_flow_item(struct gw_item_param *param,
+						struct doca_gw_match *match, uint8_t type)
+{
+	struct rte_flow_item *item = param->item;
+	struct doca_ip_addr src_ip = GET_SRC_IP(match, type);
+	struct doca_ip_addr dst_ip = GET_DST_IP(match, type);
+	struct rte_flow_item_ipv4 *spec = &param->item_param.ip4.spec;
+	struct rte_flow_item_ipv4 *mask = &param->item_param.ip4.mask;
+
+	item->type = RTE_FLOW_ITEM_TYPE_IPV4;
+	if (src_ip.a.ipv4_addr){
+		spec->hdr.src_addr = src_ip.a.ipv4_addr;
+		if (src_ip.a.ipv4_addr == 0xffffff)
+			param->flags |= (type == OUTER_IP ? SET_OUTER_SRC_IP : SET_INNER_SRC_IP);
+	}
+
+	if (dst_ip.a.ipv4_addr){
+		spec->hdr.dst_addr = dst_ip.a.ipv4_addr;
+		if (dst_ip.a.ipv4_addr == 0xffffff)
+			param->flags |= (type == OUTER_IP ? SET_OUTER_DST_IP : SET_INNER_DST_IP);
+	}
+
+	if (param->flags) {
+		item->spec = spec;
+		item->mask = mask;
+		param->apply = set_ipv4_flow_item;
+	}
+}
+
+/**
+ * @brief todo.
+ *    
+ * @param match
+ *
+ * @return 
+ */
+static int set_ipv6_flow_item(__rte_unused struct gw_item_param *param, __rte_unused struct doca_gw_match *match)
+{
+	return 1;
+}
+
+static void build_ipv6_flow_item(struct gw_item_param *param, 
+			struct doca_gw_match *match, uint8_t type)
+{
+	struct rte_flow_item *item = param->item;
+	struct doca_ip_addr src_ip = GET_SRC_IP(match, type);
+	struct doca_ip_addr dst_ip = GET_DST_IP(match, type);
+	struct rte_flow_item_ipv6 *spec = &param->item_param.ip6.spec;
+	struct rte_flow_item_ipv6 *mask = &param->item_param.ip6.mask;
+
+	item->type = RTE_FLOW_ITEM_TYPE_IPV6;
+	if (!is_ipv6_zero(src_ip.a.ipv6_addr)){
+		memcpy(spec->hdr.src_addr, src_ip.a.ipv6_addr, sizeof(src_ip.a.ipv6_addr));
+		if (is_ipv6_optional(src_ip.a.ipv6_addr))
+			param->flags |= (type == OUTER_IP ? SET_OUTER_SRC_IP : SET_INNER_SRC_IP);
+	}
+
+	if (!is_ipv6_zero(dst_ip.a.ipv6_addr)){
+		memcpy(spec->hdr.dst_addr, dst_ip.a.ipv6_addr, sizeof(dst_ip.a.ipv6_addr));
+		if (is_ipv6_optional(dst_ip.a.ipv6_addr))
+			param->flags |= (type == OUTER_IP ? SET_OUTER_DST_IP : SET_INNER_DST_IP);
+	}
+
+	if (param->flags) {
+		item->spec = spec;
+		item->mask = mask;
+		param->apply = set_ipv6_flow_item;
+	}
+}
+
+static void build_gre_flow_item(struct gw_item_param *param, struct doca_gw_match *match)
+{
+	struct rte_flow_item *item = param->item;
+	struct rte_flow_item_gre *spec = &param->item_param.gre.spec;
+	struct rte_flow_item_gre *mask = &param->item_param.gre.mask;
+
+	item->type = RTE_FLOW_ITEM_TYPE_GRE;
+	if (match->tun.gre.protocol) {
+		spec->protocol = match->tun.gre.protocol;
+		if (match->tun.gre.protocol == UINT16_MAX)
+			param->flags |= SET_GRE_PROTOCOL;
+	}
+
+	if (param->flags) {
+		item->spec = spec;
+		item->mask = mask;
+		//param->apply = set_gre_flow_item;		
```

Comment:
not that GRE you should test flags, and if key is on, you should add GRE, but let save for later.
lets go with vxlan as first use case, it is easier